Add methods to `reduce`, specialized for `merge` on Dict

jlapeyre opened the pull request:
This is discussed in #21672.
`merge` is not efficient when called with a large number of `Dict` arguments. This PR makes `merge` efficient in this case.

For example
```julia
a = [Dict(:a => 1 , :b=>2) for i in 1:2000];
@time merge(a...);  # twice for jit
```

master 46dcb35e1b92bf252aee078a90e42557b48376bc
0.262732 seconds (25.95 k allocations: 120.125 MiB, 23.43% gc time)

This commit
0.000516 seconds (13 allocations: 63.641 KiB)

This does not affect other methods for `merge` that I have found. For instance, for named tuples, and `OrderedDict` in `DataStructures`. They do not use the code that I changed.

The inefficiency is due to the performance of this kind of construction
```
promoteK(K, d, ds...) = promoteK(promote_type(K, keytype(d)), ds...)
```
I first reimplemented the type promotion using `reduce` which is efficient. But, I replaced this with code copied from `DataStructures`, which uses a simple loop, as this is compact and very transparent.

There are similar inefficiencies in other (important) methods. I have a prepared a branch fixing these in a similar way.  A more general solution, which I did not investigate, is to make the construction above efficient.

jlapeyre committed:
add method to `reduce`, specialized for `merge` (#21672)

nalimilan commented:
Thanks. Unfortunately, AFAICT `merge!` is no longer inferrable with these changes (you can check that with `@code_warntype`).

I'm concerned that this change amounts to optimizing for a situation which should not really happen in the first place. Splatting shouldn't be used when the number of arguments is large. How about implementing the special `reduce` method discussed at #21672 instead?

jlapeyre commented:
That's strange, my change should not affect `merge!` at all.

nalimilan commented:
Sorry, I meant `merge`, not `merge!`.

jlapeyre commented:
Ok. Yes, that's not an acceptable loss. I see that  `merge` for OrderedDict is also not inferrable.

I'll look into it.

jlapeyre commented:
I did not find a solution that is both efficient and inferable. The generic method for `reduce(merge, array_of_dicts)` is already reasonably fast and light on resources. However, I did write another method for `reduce` that reduces the time by 30% for some tests.  I will change the pull request. Is this as simple as the suggestion from github ? :
> Add more commits by pushing to the gjl/merge branch on jlapeyre/julia

Will the commit for the PR be updated, or do I need to do something more ? Of course, it would be best if the changes from the previous commit are not visible. 

To be useful, the method for `reduce` with `merge` has to take an array of type `Any` containing `Dicts` of different types. I guess there is no way  to write an inferable method with this constraint.  (I did not look into the possibility of a `Tuple` of `Dicts`.)

nalimilan commented:
> I did not find a solution that is both efficient and inferable. The generic method for reduce(merge, array_of_dicts) is already reasonably fast and light on resources. However, I did write another method for reduce that reduces the time by 30% for some tests. I will change the pull request. Is this as simple as the suggestion from github ? :
> 
>     Add more commits by pushing to the gjl/merge branch on jlapeyre/julia
>
> Will the commit for the PR be updated, or do I need to do something more ? Of course, it would be best if the changes from the previous commit are not visible.

Yes, that's it. You can amend the previous commit and force-push if you want to erase it.


> To be useful, the method for reduce with merge has to take an array of type Any containing Dicts of different types. I guess there is no way to write an inferable method with this constraint. (I did not look into the possibility of a Tuple of Dicts.)

Actually, the method can accept any array, and if the array type is concrete (e.g. `Vector{Dict{Symbol,Int}}`), then the return type is inferrable.

jlapeyre commented:
> Actually, the method can accept any array, and if the array type is concrete (e.g. Vector{Dict{Symbol,Int}}), then the return type is inferrable.

Yes, it should do the right thing for your example. I have been concentrating on the non-trivial cases. I just meant that it it ought to do something better than  return `Dict{Any,Any}` if the input is `Vector{Any}`.  I have been looking at discussions of how to handle type promotion or inference in various cases, and it left my head spinning. It's hard to  predict the typical use case. In the case that lead to this, I had an array `Vector{Any}` all of whose elements were of the same type: concretely typed `Dicts`.  Using the approach of `map` would be best in this case. That is, start with an output `Dict` of the same type as that of the first element, and widen and copy if necessary. On the other hand, I know that in some cases a `Dict` with certain concrete key and value types is not more efficient than one with more general types. But, I don't know the details.  There is even precedent for making `reduce` return the best container. That is, don't look at the keytype, but rather the type of each key.  This is kind of like what `[x for x in array]` does. 

Instead, I have been following the example of `merge`, that is compute the correct output type once, before doing the merge.

nalimilan commented:
Makes sense. I think you can just use `Dict{mapreduce(keytype, promote_type, vec_of_dicts), mapreduce(valtype, promote_type, vec_of_dicts)}` to compute the type of the `Dict` to return. When `vec_of_dicts` has a concrete type, inference is able to figure out the return type, but it also works when it's not concrete. So that's the best of both worlds.

jlapeyre commented:
I'm questioning the value of continuing this PR. The arguments against continuing are

1.  The existing method for `reduce(merge,arr_of_dicts)` is not inefficient in an algorithmic sense (AFAICT).  So far, I can only  decrease the execution time by roughly a constant factor, independent of the number of dictionaries. In most cases that I tried, the decrease in time is  30% or so. (But,  for `OrderedDicts` it may be as much as 90%.)

2. I compare a couple of new methods and the default `reduce` method for a variety of inputs. Which method is fastest depends strongly on the types of dictionaries and the type of container. I get the biggest gain in performance for `OrderedDicts`, roughly  a factor of 10.  In some cases, the existing method is faster even though it is more naive and does far more allocation.

3. Because of ongoing major changes in the iteration protocol and in optimization, the best method for the various cases is likely to change before v0.7 or v1.0 is released.

4. Because of item number 2, the logic for implementing the new methods adds a bit of complexity and fragility that must be weighed against the benefits. 

In short, this may be premature optimization.  But, I am willing to consider arguments for proceeding with the PR.

nalimilan commented:
AFAICT the time increases more than linearly, because `merge` allocates a new dict at each step, which needs to include all contents treated in all previous iterations. And indeed, in a simple benchmark doubling the number of dicts more than doubles the time. Below you can see that multiplying the number of dicts by 256 multiplies the time by... 27,000!
```julia
julia> a = [Dict(rand()=>i for i in 1:1000) for j in 1:512];

julia> for i in 1:9; @btime reduce(merge, $(a[1:2^i])); end
  32.840 μs (6 allocations: 68.42 KiB)
  183.019 μs (25 allocations: 681.52 KiB)
  432.450 μs (53 allocations: 1.73 MiB)
  1.588 ms (115 allocations: 8.90 MiB)
  4.604 ms (227 allocations: 25.91 MiB)
  25.088 ms (457 allocations: 127.92 MiB)
  87.035 ms (905 allocations: 399.95 MiB)
  256.262 ms (1807 allocations: 1.27 GiB)
  874.068 ms (3605 allocations: 4.75 GiB)

# Merging in-place is much faster
julia> @btime reduce(merge!, $(copy(a[1])), a)
  44.914 ms (0 allocations: 0 bytes)
```

That said, if you don't see the need for this, there's no hurry to implement this specialized method. The issue in which I mentioned this was about vectors of vectors, which are much more common and cannot be concatenated in-place as easily as dicts.

jlapeyre commented:
I see your point. In number 2 in my last post I neglected to mention that efficiency also varies with the content of the dictionaries.  I have been testing mostly with many dictionaries and a small set of discrete keys and values, because that was the way the problem arose in an application.

jlapeyre commented:
I changed the content and title of the PR.  This PR implements methods for `reduce` with operator `merge`. 

I tested this on several cases including the one given in https://github.com/JuliaLang/julia/pull/26440#issuecomment-373958898.

jlapeyre commented:
The test suite fails. This PR introduces a method ambiguity: `reduce(f, S::SharedArrays.SharedArray)`. This is a common problem. Is there a better way to solve this than to add a wrapper method for each ambiguity ?

nalimilan commented:
Unfortunately not. Though you could implement only the method for `AbstractVector{<:AbstractDict}`. The drawback would be that people storing their dicts in a `Vector{Any}` wouldn't benefit from the optimization. But that would make the code simpler and more robust, so maybe that's a good idea idea. For example, with the current state of the PR if the first element in the vector is a dict but others are not, the code is going to throw an error about `keytype` while a valid `merge` method could work.

jlapeyre commented:
That's the price for multiple dispatch. 

Disallowing `Vector{Any}` is not very palatable. It would be a more complicated API for a user who needs the optimized version. For instance, `JSON.parse` returns a `Vector{Any}` of dicts, even if they are all of the same type.

> For example, with the current state of the PR if the first element in the vector is a dict but others are not, the code is going to throw an error about keytype while a valid merge method could work.

Yes, for instance, if the second element is a `NamedTuple`. But in this case there is no applicable method for `merge` either. And  `reduce(merge, [dict,namedtuple])` throws an error in the master branch as well. Nor is there a method for `merge` that merges a dict with an array of pairs.

nalimilan commented:
> Yes, for instance, if the second element is a NamedTuple. But in this case there is no applicable method for merge either. And reduce(merge, [dict,namedtuple]) throws an error in the master branch as well. Nor is there a method for merge that merges a dict with an array of pairs.

That's the case for named tuples, but any package could create a custom type which is not an `AbstractDict` and yet supports `merge` with a `Dict`. Granted, that's not completely standard, but it should be possible. To be safe, the specialized `reduce` method should check that all inputs are `AbstractDict`, and if no fall back to the generic method.

jlapeyre commented:
Good point. The optimized version in the current PR would cut off that opportunity. I'll work on fixing that.

jlapeyre commented:
I edited the PR to check that the types of all elements in an array of type `Vector{Any}` are subtypes of `Dict`. This required the addition of just [one line](https://github.com/JuliaLang/julia/pull/26440/files#diff-1fe99d7ee27f2825eed089ccbd397effR370)

The generic `reduce` method calls a chain of methods. If one of the items to be merged is not a `Dict`, then I call the first method in that chain, rather than using the method that is specialized for `Dict`s.  IIRC, there is a way to call `reduce` while specifying that I do not want to specialize on the first argument. But, I don't recall how this works.

nalimilan commented on base/abstractdict.jl:
```diff
@@ -335,6 +335,75 @@ function _typeddict(d::AbstractDict, others::AbstractDict...)
+    return dto
+end
+
+function _merge!(dto::AbstractDict,  dfrom::AbstractDict)
```
Isn't this just `merge!`?

jlapeyre replied:
Probably. In any case, it is never called. It should be removed.

nalimilan replied:
I don't understand. `merge!` takes splatted arguments, but that's for multiple *dictionaries*, not for multiple *pairs*. Here the method only allows two arguments, so it's equivalent to `merge!`.

jlapeyre replied:
Neither do I ! :) I wasn't thinking of pairs. I'll state how I see it and see if you or agree or see a mistake.
   1. We are talking about the methods `merge!(d::AbstractDict, others::AbstractDict...)` and `_merge!(dto::AbstractDict,  dfrom::AbstractDict)`.
    2. For `Dict`s `a` and `b`, there is no observable difference between `merge!(a,b)` and `_merge!(a,b)`
    3. For the call `merge!(a,b)`, the argument `b` appears in the body in a `Tuple` of one element. The routine must iterate over this tuple.  In `_merge!(a,b)`, there is no tuple and no corresponding iteration. So the latter may be faster than the former.

I did some benchmarks and `_merge!` is indeed faster. How much faster depends of course on the characteristics of the input dicts.  So you gain a bit of efficiency at the expense of writing another (short) method. One can argue for or against `_merge!(a,b)` by arguing that the efficiency outweighs the code complexity, or vice-versa.

jlapeyre replied:
A concrete example:
```
julia> print(d1)
Dict(7=>9,9=>5,4=>3,2=>8,3=>5,5=>6,1=>1)
julia> print(d2)
Dict(7=>6,9=>6,2=>3,3=>10,5=>9,8=>10,6=>2)
julia> @btime merge!($(copy(d1)),d2);
  133.953 ns (0 allocations: 0 bytes)

julia> @btime _merge!($(copy(d1)),d2);
  87.529 ns (0 allocations: 0 bytes)
```

nalimilan replied:
Not sure whether the performance difference is significant given how short these timings are, but if it makes a significant difference in your concrete use case then it's worth filing a separate issue. If the difference is real a two-argument method could be added to `merge!`.

jlapeyre replied:
Well, the ratio is significant, and I think that's more important. But, the point is moot. I just wrote a discourse post about the general question. After thinking while composing the questions,  I'm quite comfortable with implementing all your suggestions. I think its the better choice. As you say, maybe further optimizations can be made in the future.

nalimilan commented on base/abstractdict.jl:
```diff
@@ -335,6 +335,75 @@ function _typeddict(d::AbstractDict, others::AbstractDict...)
+    return dto
+end
+
+function _merge!(dto::AbstractDict, dicts::AbstractArray)
```
This can be simplified to `_merge!(dto, dicts[1], dicts[2:end])`.

jlapeyre replied:
Yes. But, shouldn't the last argument be a `view` ? I did benchmark the other use of `view` in this PR and the increased efficiency was not insignificant.

nalimilan replied:
Yeah, a view might be slightly better.

nalimilan commented on base/abstractdict.jl:
```diff
@@ -335,6 +335,75 @@ function _typeddict(d::AbstractDict, others::AbstractDict...)
+
+function reduce(m::typeof(merge), items::AbstractVector)
+    isempty(items) && throw(ArgumentError("Cannot merge empty vector"))
+    length(items) == 1 && return copy(first(items))
```
` reduce(merge, [x])` currently returns `x` without copying, so this isn't needed: you can pass an empty array when there's a single element.

Same below.

jlapeyre replied:
I don't understand `empty array`.  To agree with the current behavior, we could return `first(items)`. But, you are saying something else ?

nalimilan replied:
I mean that `reduce(merge, items)` is equivalent to `first(items)` when there's only one element, so no need to add special code for that case.

jlapeyre replied:
Ok. Got it. Simply remove the line and there is no observable difference. You are arguing again for simplicity over efficiency.

nalimilan commented on base/abstractdict.jl:
```diff
@@ -335,6 +335,75 @@ function _typeddict(d::AbstractDict, others::AbstractDict...)
+    length(items) == 1 && return copy(first(items))
+    i1 = firstindex(items) + 1
+    i2 = lastindex(items)
+    return reduce(m,first(items), view(items,i1:i2))
```
`i2` can be replaced with `end`.

jlapeyre replied:
Good. I wasn't sure about it.

jlapeyre replied:
Are you sure about replacing `i2` with `end`?
```
julia> a = collect(1:10);
julia> view(a, 1:end);
ERROR: syntax: missing last argument in "1:" range expression 
julia> view(a, 1:lastindex(a));

julia> ex = Expr(:call,:getindex, :a, Expr(:call,:(:),1,:end))
:(getindex(a, 1:end))
julia> eval(ex)
ERROR: UndefVarError: end not defined
```
Now I recall having tried `end` first. Must be a good reason for this.

jlapeyre replied:
Thanks. That will work. Anyway the method in question has been removed.
I just rebased and pushed all the changes in your last round of comments.

nalimilan reviewed (commented):
Thanks. I'm a bit worried about the number of new methods, which apart from duplicating a lot of features can also introduce ambiguities. A possible approach would be this:
1. Have `reduce(m::typeof(merge), items::AbstractVector)` call `_merge!(reduce(typejoin, items), items::AbstractVector)`.
2. Have `_merge!(::Type{<:AbstractDict}, items::AbstractVector)` call `reduce(merge!, _typeddictA(items), items)`, and a fallback `_merge!(::Type, items::AbstractVector)` call `reduce(merge, first(items), view(items, firstindex(items)+1:end))`.

This system would allow other types to define optimized methods without introducing ambiguities.

jlapeyre commented:
Thanks for reviewing. I agree with your ideas for reducing the number of methods and ambiguity. But, In point 1.,  shouldn't  `reduce(typejoin,items)` be `mapreduce(typeof,typejoin,items)` ?

What's the most efficient way to make the changes ? Do you want to do it ? It's easy for me to make the changes and rebase  once more.

jlapeyre commented:
I did not refresh and see your new comments. I'll read them now...

nalimilan commented:
You haven't removed the methods as I suggested in my review. Have you tried that?

jlapeyre commented:
I thought I did remove them. Maybe, I misunderstood the comments and did make all the changes.
I did implement points 1 and 2 [in in this comment](https://github.com/JuliaLang/julia/pull/26440#pullrequestreview-110787710), and these replaced other methods.

I see now that I did forget to replace `_merge!(dto::AbstractDict, dfrom::AbstractDict)`  with `merge!`.

jlapeyre commented on base/abstractdict.jl:
```diff
@@ -335,6 +335,60 @@ function _typeddict(d::AbstractDict, others::AbstractDict...)
+end
+
+function reduce(::typeof(merge), d::T, dicts::AbstractVector{T}) where T <: AbstractDict{K,V} where {K,V}
+    isempty(dicts) && return copy(d)
```
This can probably be removed. It would be consistent with your last review.

nalimilan commented:
My proposal was to drop the specialized `reduce` methods for `AbstractArray{<:AbstractDict}`, and handle everything in the generic `reduce(::typeof(merge), ...)` methods instead.

jlapeyre commented:
OK, I'll look at this today, if I can get tuit, else the weekend.

jlapeyre commented:
I think reducing the *entire* PR to this
```
reduce(m::typeof(merge), items::AbstractVector) = _merge!(mapreduce(typeof,typejoin,items), items)
_merge!(::Type{T}, items::AbstractVector) where T <: AbstractDict = reduce(merge!,T(),items)
_merge!(::Type, items::AbstractVector) = reduce(merge,first(items),@view items[firstindex(items)+1:end])
```
is consistent with what you want, right ?   I tested it with your example and those I have been using. It performs well compared to the more specialized versions, maybe even better in some cases.  And `reduce(merge, vector_of_named_tuples)`  works.

fredrikekre commented on stdlib/SharedArrays/src/SharedArrays.jl:
```diff
@@ -559,6 +559,8 @@ similar(S::SharedArray) = similar(S.s, eltype(S), size(S))
     mapreduce(fetch, f, Any[ @spawnat p reduce(f, S.loc_subarr_1d) for p in procs(S) ])
 
+reduce(::typeof(merge), S::SharedArray) =
+    throw(ArgumentError("cannot call `merge` on a `SharedArray`"))
```
Why is this needed at all?

nalimilan replied:
I guess because of the ambiguity with the previous method?

fredrikekre replied:
So this PR breaks `reduce(merge, ::SharedArray)`?

nalimilan replied:
Ah, right, better change it to:
```julia
reduce(::typeof(merge), S::SharedVector) = invoke(reduce, Tuple{Any,SharedArray}, merge, S)
```
(I know nothing about `SharedArray` so I don't know whether the new optimized method would work or not.)

jlapeyre replied:
*  `@view`: Ok.

 * `MethodError`: Yes, this is better

 * Ah `invoke` (!) That's the function I mentioned whose name I could not remember. I just threw an error instead, because  "type of SharedArray elements must be bits types".  The chance that a method is written for this case near zero. But, of course it should be allowed.

* Tests: I will add tests. I doubt the code would be covered otherwise.

* Commas: It's my preference to always add a space. I was trying to follow the example of some functions that I saw nearby in the file (IIRC). The insertion of a space or not after a comma in "base/abstractdict.jl" is more or less random. But, sure, I'll do it.

jlapeyre commented:
I discovered a bug in the latest version of this PR. The constructor `T()` may fail sometimes. Like this

```julia
julia> similar(Dict{T,V} where {T,V})
Dict{Any,Any} with 0 entries
julia> similar(Dict{T,Int} where T)
ERROR: MethodError: Cannot `convert` an object of type Tuple{} to an object of type Dict{T,Int64} where T
```

jlapeyre commented:
Here is the previous version of the PR, with a bug fix. The type of the  output `Dict` must be concrete, of course.  But this was not always the case in the previous attempt.

I added a few tests.

jlapeyre commented:
Can someone comment on the status of this PR please ?  The CI messages above show failure. But, if I recall correctly, this has happened before when there were no problems on my end.  If it is on my end, I can try to track down the problem.

nalimilan commented:
I'm afraid the failure is due to this PR. Looks like it's due to using `Dict` in a file in abstractdict.jl, where it's not yet defined.

jlapeyre commented:
Of course, there it is in the CI output. Don't know how I missed that.

jlapeyre commented:
Since I rebased and there is no record in the git log, I'll mention it here: I just fixed this:

> Looks like it's due to using Dict in a file in abstractdict.jl, where it's not yet defined.

nalimilan commented on base/abstractdict.jl:
```diff
@@ -335,6 +335,10 @@ function _typeddict(d::AbstractDict, others::AbstractDict...)
 end
 
+reduce(m::typeof(merge), items::AbstractVector) = _merge!(mapreduce(typeof, typejoin, items), items)
+_toany(T) = isconcretetype(T) ? T : Any
```
This function isn't used in this file, better move it to the other one. But is it really needed? The test which mentions `Any` vs. `Real` doesn't sound correct since one of the dicts uses `Symbol` keys.

Actually, it seems what is needed is not to use `K` and `V` at all (i.e. do `::Type{<:Dict}`), but instead compute them with `mapreduce(keytype, promote_type, items)` and `mapreduce(valuetype, promote_type, items)`.

jlapeyre replied:
Agreed it should be moved. And, I agree it is not tested properly. If I recall, the performance with, say `Real` rather than `Any` was much worse. But, that can be optimized later, and may be better or worse by now with changes to v0.7 in the last month. The main reason is that I think `Any` is better than a more specific abstract type. There must be some discussion on this kind of thing. It is not mentioned explicitly in the [style guide](https://docs.julialang.org/en/latest/manual/style-guide/). But it  is similar in spirit to the admonition against weird `Union` types.

What do you think about moving all the new code to `dict.jl` ? The new lines in `abstractdict.jl` don't mention `AbstractDict` or subtypes thereof at all.

> Actually, it seems what is needed is not to use K and V at all (i.e. do ::Type{<:Dict}), but instead compute them with 

I think you are suggesting to use `mapreduce` to compute each of the key and value types rather than the type of `Dict` at one go ?  This sounds ok. But, that still may result in `Dict{A,Real}` etc. It would be nice to get some feedback or testing in the wild before it is frozen in 1.0.

nalimilan replied:
> Agreed it should be moved. And, I agree it is not tested properly. If I recall, the performance with, say Real rather than Any was much worse. But, that can be optimized later, and may be better or worse by now with changes to v0.7 in the last month. The main reason is that I think Any is better than a more specific abstract type. There must be some discussion on this kind of thing. It is not mentioned explicitly in the style guide. But it is similar in spirit to the admonition against weird Union types.

I don't think we need to think about what's the best type here: it must be the same as what `merge(::Dict, ::Dict)` gives, that's all.



> What do you think about moving all the new code to dict.jl ? The new lines in abstractdict.jl don't mention AbstractDict or subtypes thereof at all.

Yeah, on the other hand `merge` is defined here to it makes sense to define `reduce(::typeof(merge), ...)` here too.

> I think you are suggesting to use mapreduce to compute each of the key and value types rather than the type of Dict at one go ? This sounds ok. But, that still may result in Dict{A,Real} etc. It would be nice to get some feedback or testing in the wild before it is frozen in 1.0.

Yes, but AFAIK that's what `merge` does already, so that what we should do. If we want to change that, better file a separate issue.

nalimilan commented on base/abstractdict.jl:
```diff
@@ -335,6 +335,10 @@ function _typeddict(d::AbstractDict, others::AbstractDict...)
     Dict{K,V}(d)
 end
 
+reduce(m::typeof(merge), items::AbstractVector) = _merge!(mapreduce(typeof, typejoin, items), items)
```
`m` isn't a typical name for this argument, and it's not needed at all.

jlapeyre replied:
Agreed.

nalimilan commented on base/dict.jl:
```diff
@@ -684,6 +684,9 @@ end
 
 filter!(f, d::Dict) = filter_in_one_pass!(f, d)
 
+## Another method for this function is in abstractdict.jl.
```
Even better: "the fallback method". Also, a single # is enough.

jlapeyre replied:
Yes, a single # is consistent throughout the file.

nalimilan commented on test/dict.jl:
```diff
@@ -876,6 +876,18 @@ end
+    @test dr == Dict(1=>2, 3=>4)
+    @test typeof(dr) == Dict{Int,Int}
+    # The key type of the returned Dict should be `Any` rather than `Real`
+    @test typeof(reduce(merge, [d2, d3])) == Dict{Any,Int}
```
Also test the value of the result.

jlapeyre replied:
Ok.

nalimilan commented on base/abstractdict.jl:
```diff
@@ -335,6 +335,10 @@ function _typeddict(d::AbstractDict, others::AbstractDict...)
 
+reduce(m::typeof(merge), items::AbstractVector) = _merge!(mapreduce(typeof, typejoin, items), items)
+_toany(T) = isconcretetype(T) ? T : Any
+_merge!(::Type, items::AbstractVector) = reduce(merge, first(items), view(items,firstindex(items)+1:lastindex(items)))
```
Break line before `reduce` to keep below 92 chars limit.

jlapeyre replied:
> Break line
Ok.

> the second argument to reduce is only used for empty collections

Yes, that's usually correct. I don't recall if I innovated here or was copying something. But, it may be that `invoke` is more robust in any case.

nalimilan reviewed (commented):
Cool. I have a few more comments.

Something that occurred to me is that `reduce` is a weird chose for this operation if keys can appear in several input dicts: in that case, it's undefined which key will be used. So `foldl` and `foldr` will have to be used when you want to keep the first or last occurrence. These can always be implemented later, though.

jlapeyre commented:
Just noticed you are writing. Thanks for the review. I will read your comments.

I am looking at the build logs, trying to see if there is a problem. I'm not familiar with what not-to-worry output looks like... but for failed builds and even for a linux build that "passes", `grep rawlog.txt test | grep jl  | wc` gives 12 lines. Most of them don't look suspicious. The only two that are particular to testing features are
```
      From worker 3:	macro expansion at /tmp/julia/share/julia/test/dict.jl:906 [inlined]
      From worker 3:	top-level scope at /tmp/julia/share/julia/test/dict.jl:898
```
Those lines are in a test set that is 2 test sets below the one I added.

nalimilan commented:
Good catch. That's probably due to the problem I pointed [here](https://github.com/JuliaLang/julia/pull/26440#discussion_r189552092), i.e. `merge` is broken for named tuples.

I'm starting to wonder whether we shouldn't have just added a method for `reduce(::typeof(merge), AbstractVector{<:Dict})` or even `reduce(::typeof(merge), Vector{<:Dict})`. That would have been simpler.

jlapeyre commented:
> Something that occurred to me is that reduce is a weird chose for this operation if keys can appear in several input dicts

Yes, I did not address this.  This existing method `merge(combine, d::AbstractDict, others::AbstractDict...)` gives some control over handling duplicate keys. This is not supported by this solution.

jlapeyre commented:
> I'm starting to wonder whether we shouldn't have just added a method for reduce(::typeof(merge), AbstractVector{<:Dict}) or even reduce(::typeof(merge), Vector{<:Dict}). That would have been simpler.

I think there are common enough use cases where the `Dict`s are in a `Vector{Any}`.  IIRC I encountered this  using `JSON.jl`. I read a file and got a `Vector{Any}` with `Dict`s.  But, maybe it reasonable to require that the user convert to `Vector{<:Dict}` in order to take advantage of the efficient implementation.

jlapeyre commented:
The bug that shows up in this PR is due to the test code, not the changes to base.
The following triggers the bug in the master branch from yesterday, i.e. *without* the current PR. 
```
reduce(merge,[(a=1,b=2),(c=3,d=4)])
```
From your comment,
> merge is broken for named tuples

its not clear to me if you already knew this.
The following is apparently ok
```
reduce(merge, ((a=1,b=2),(c=3,d=4)))
```
As is this,
```
merge((a=1,b=2),(c=3,d=4))
```

nalimilan commented:
OK, then it's not due to what I suspected. Anyway better drop this test for now.

jlapeyre commented:
I just saw your last responses. I didn't realize I have to click on the diff to see all of the comments. Anyway, I realized in the meantime what you wrote there: The behavior of all three of `merge(a...)`, and the unoptimized `reduce(merge,...`, and the optimized one should be identical.

Here is a proposal. (`reduce1` would be replace by `reduce`)

```julia
using Test

reduce1(::typeof(merge), items::AbstractVector) =
    _merge!(mapreduce(typeof, typejoin, items), items)

_merge!(::Type, items::AbstractVector) =
    invoke(reduce, Tuple{Any,AbstractArray}, merge, items)

_merge!(::Type{Dict{K,V}}, items::AbstractVector) where {K,V} =
    reduce(merge!, Dict{K,V}(), items)

function testone(vin,dout)
    @test reduce1(merge, vin) == dout
    @test reduce(merge, vin) == dout    
    @test merge(vin...) == dout
end

function mytests()
    testone([Dict( 1 => 2), Dict( 1.0 => 2.0)],
            Dict(1.0=>2.0)),

    testone([(a=1,b=2), (a=1,b=3,c=4)],
            (a = 1, b = 3, c = 4)),

    testone([Dict(1=>2), Dict(2=> 1.0 + im)],
            Dict(2=>1.0+1.0im,1=>2.0+0.0im)),

    testone([Dict(1=>2), Dict(3=>4)],
            Dict(3=>4,1=>2)),

    testone([Dict(3=>4), Dict(:a=>5)],
            Dict( :a => 5, 3 => 4))

end
```
This gives
```julia
julia> mytests()
(Test Passed, Test Passed, Test Passed, Test Passed, Test Passed)
```

It also allows one to write optimized methods for `_merge!` for `NamedTuple`,  etc.

nalimilan commented:
Looks good, except for `_merge!(::Type{Dict{K,V}}, ...`, see my comment above about `promote_type`. Tests should also check that types are equal to what you get using `merge`.

jlapeyre commented:
> Tests should also check that types are equal to what you get using merge.

Ok, I erroneously thought that  `==` cares about container type.

Regarding `promote_type`, I'm not sure what you mean.... Maybe this ?
```julia
function reduce1(::typeof(merge), items::Vector{<:Dict})
    K = mapreduce(keytype, promote_type, items)
    V = mapreduce(valtype, promote_type, items)
    return reduce(merge!, Dict{K,V}(), items)
end
```

Also, for the same tests, this works as well
```julia
reduce1(::typeof(merge), items::AbstractVector) =
     _merge!(mapreduce(typeof, promote_type, items), items)
```
That is, with `promote_type` rather than `type_join`

jlapeyre commented:
Is there a way to run this build again ? It seems likely that the failure was not due to the PR.

StefanKarpinski commented:
If you rebase and force push the branch, CI will run again.

nalimilan commented:
Bump. Can you rebase?

jlapeyre commented:
Yes. But, I have a deadline to meet tomorrow. Thanks

jlapeyre commented:
I'm not sure of what I'm doing.  I tried the following, which did nothing in the end. But, I can see that `fetch` is fetching from my fork of julia master. My fork has not been updated
```
lapeyre@ribot~/j/f/julia> git branch
* gjl/merge
  master
lapeyre@ribot~/j/f/julia> git fetch origin master
From github.com:jlapeyre/julia
 * branch                  master     -> FETCH_HEAD
lapeyre@ribot~/j/f/julia> git rebase --interactive origin/master   # I do not edit, just save the file that is opened
Successfully rebased and updated refs/heads/gjl/merge.
lapeyre@ribot~/j/f/julia> git push --force origin gjl/merge
Everything up-to-date
```

So, I try this:
```
lapeyre@ribot~/j/f/julia> git remote -v 
origin	git@github.com:jlapeyre/julia.git (fetch)
origin	git@github.com:jlapeyre/julia.git (push)
upstream	https://github.com/JuliaLang/julia (fetch)
upstream	https://github.com/JuliaLang/julia (push)
lapeyre@ribot~/j/f/julia> git fetch upstream  master 
remote: Counting objects: 7157, done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 7157 (delta 3010), reused 3010 (delta 3010), pack-reused 4145
Receiving objects: 100% (7157/7157), 5.21 MiB | 9.94 MiB/s, done.
Resolving deltas: 100% (5382/5382), completed with 851 local objects.
From https://github.com/JuliaLang/julia
 * branch                  master     -> FETCH_HEAD
   a96d847592..9a4ecaa4e4  master     -> upstream/master
```

Any ideas about what to do next ?

jlapeyre commented:
This test added in this PR test fails
```
/tmp/julia/share/julia/test/dict.jl:902
  Got exception ErrorException("`reduce(op, v0, itr)` is deprecated, use `reduce(op, itr; init=v0)` instead") outside of a @test
  `reduce(op, v0, itr)` is deprecated, use `reduce(op, itr; init=v0)` instead
```
I'll fix this and rebase.

nalimilan committed:
Use standard function syntax and more explicit name

rfourquet commented on test/dict.jl:
```diff
@@ -900,7 +900,7 @@ end
 
 @testset "Dict reduce merge" begin
-    f = (i::Vector{<:Dict}, o) -> begin
+    function f(i::Vector{<:Dict}, o)
```
Should be `check_merge` ?

jlapeyre replied:
I agree `check_merge` is better. I modeled the name `f` on a similar function in the file.

nalimilan committed:
Fix function name

AzamatB commented:
Bump

jlapeyre commented:
I'ts been a while, but I think there is nothing for me to do at this point.
But, reviewing the comments it looks like nanosoldier was never run. Maybe worth checking ?

nalimilan commented:
Nanosoldier will only be useful if there are benchmarks for `reduce(merge, dicts)`, which is probably not the case. It would be nice to add such benchmarks to BaseBenchmarks, though. Anyway, can you just show one or two benchmarks comparing with master?

nalimilan committed:
Merge branch 'master' into gjl/merge

nalimilan commented:
I'll merge this tomorrow if nobody objects, as I don't want it to miss the release again.

rfourquet commented on base/dict.jl:
```diff
@@ -691,6 +691,12 @@ end
 
 filter!(f, d::Dict) = filter_in_one_pass!(f, d)
 
+function reduce(::typeof(merge), items::Vector{<:Dict})
```
Any reason not to allow `AbstractVector` ?
(sorry for a last minute comment!)

nalimilan replied:
IIRC it triggered ambiguities (that's probably buried in a comment above). It would be useful to try to generalize, but for now I'd say it's already an improvement.

nalimilan commented:
Thanks @jlapeyre! Sorry it took so long...
